src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
I think there you could easily prevent that by having the Pipe maintain the knowledge that a transaction is still ongoing.

As you already said it mimics a socket, so any single connection (instance of NamedPipe) is independent of another. At least if I understood your previous comment correctly the following is allowed (and independent):

```
NamedPipe pipe1 = new SMBClient().connect(...).authenticate(...).connectTree(pipe);
NamedPipe pipe2 = new SMBClient().connect(...).authenticate(...).connectTree(pipe);
pipe1.transact(..);
pipe2.transact(..);
```

Right?